src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerJsonProcessor.java

```diff
@@ -239,6 +243,94 @@ public Content buildProviderJson(final Repository repository,
         ContentTypes.APPLICATION_JSON));
   }
 
+  /**
+   * Merges an incoming set of packages.json files.
+   */
+  public Content mergePackagesJson(final Repository repository, final List<Payload> payloads) throws IOException {
+    Set<String> names = new HashSet<>();
+    for (Payload payload : payloads) {
+      Map<String, Object> json = parseJson(payload);
+      Map<String, Object> providers = (Map<String, Object>) json.get(PROVIDERS_KEY);
+      names.addAll(providers.keySet());
+    }
+    return buildPackagesJson(repository, names);
+  }
+
+  /**
+   * Merges incoming provider JSON files, producing a merged file containing only the minimal subset of fields that we
+   * need to download artifacts.
+   */
+  public Content mergeProviderJson(final Repository repository, final List<Payload> payloads, final DateTime now)
+      throws IOException
+  {
+    String time = now.withZone(DateTimeZone.UTC).toString(timeFormatter);
+
+    // TODO: Make this more robust, right now it makes a lot of assumptions and doesn't deal with bad things well,
+    // can probably consolidate this with the handling for rewrites for proxy (or at least make it more rational).
+    Map<String, Map<String, Object>> packages = new LinkedHashMap<>();
+    for (Payload payload : payloads) {
+      Map<String, Object> json = parseJson(payload);
+      if (json.get(PACKAGES_KEY) instanceof Map) {
+
+        Map<String, Object> packagesMap = (Map<String, Object>) json.get(PACKAGES_KEY);
+        for (String packageName : packagesMap.keySet()) {
+
+          Map<String, Object> packageVersions = (Map<String, Object>) packagesMap.get(packageName);
+          for (String packageVersion : packageVersions.keySet()) {
+
+            Map<String, Object> versionInfo = (Map<String, Object>) packageVersions.get(packageVersion);
+            Map<String, Object> distInfo = (Map<String, Object>) versionInfo.get(DIST_KEY);
+            if (distInfo == null) {
+              continue;
+            }
+
+            Map<String, Object> newDistInfo = new LinkedHashMap<>();
+            newDistInfo.put(URL_KEY, String
+                .format(REWRITE_URL, repository.getUrl(), packageName, packageVersion, packageName.replace('/', '-'),
```

Comment:
Minor: Although simple this would make a good util method to make refactoring easier in the future.

Reply:
> Minor: Although simple this would make a good util method to make refactoring easier in the future.

Related change in https://github.com/sonatype-nexus-community/nexus-repository-composer/pull/14/commits/88f058a2d3768abcb070e2f92929139d2c6ea644 (also makes the output between proxy, hosted, and group more consistent for dist portions).